### PR TITLE
[LTS] Add custom consumers with TopicManager

### PR DESCRIPTION
**Pull request by monassfaro**

### CODE REVIEW

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    private final TopicManagerService topicManagerService;
+
+    public void startDynamicConsumers(String tenantKey) {
+        Map<String, ConsumerHolder> tenantConsumerHolders = dynamicConsumersByTenant.computeIfAbsent(tenantKey, (key) -> new ConcurrentHashMap<>());
+        List<DynamicConsumer> dynamicConsumers = getDynamicConsumersByTenant(tenantKey);
+
+        dynamicConsumers.forEach(it -> topicManagerService.startNewConsumer(tenantKey, it.getConfig(), tenantConsumerHolders, it.getMessageHandler()));
+    }
+
+    public void refreshDynamicConsumers(String tenantKey) {
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java:30]: we need additional method that refresh all DynamicConsumers consumers refresh

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.icthh.xm.commons.topic.service;
+
+import com.icthh.xm.commons.topic.domain.DynamicConsumer;
+
+import java.util.List;
+
+public interface DynamicConsumerConfiguration {
+    List<DynamicConsumer> getDynamicConsumers(String tenantKey);
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfiguration.java:8]: this interface must be without tenantKey

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class DynamicConsumerConfigurationService {
+
+    private final Map<String, Map<String, ConsumerHolder>> dynamicConsumersByTenant = new ConcurrentHashMap<>();
+    private final List<DynamicConsumerConfiguration> dynamicConsumerConfigurations;
+    private final TopicManagerService topicManagerService;
+
+    public void startDynamicConsumers(String tenantKey) {
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java:23]: who will call this method on startup?

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java (hunk 1 above)
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java:30]: we heed method that refresh all

**Comment by monassfaro** [comment]
pls increase project version in gradle.version

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java (hunk 3 above)
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java:23]: TopicConfigurationService.refreshConfig

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java (hunk 1 above)
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java:37]: added

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        //start and update consumers
+        forUpdate.forEach(topicConfig -> topicManagerService.processTopicConfig(tenantKey, topicConfig, existingConsumers));
+
+        //remove old consumers
+        topicManagerService.removeOldConsumers(tenantKey, forUpdate, existingConsumers);
+
+        tenantTopicConsumers.put(tenantKey, existingConsumers);
+
+        dynamicConsumerConfigurationService.startDynamicConsumers(tenantKey);
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:85]: why dynamic consumers recreates on config refresh?

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            refreshConfig(configKey, configValue);
+        }
+    }
+
+    private void refreshConfig(String updatedKey, String config) {
+        String tenantKey = extractTenant(updatedKey);
+        Map<String, ConsumerHolder> existingConsumers = getTenantConsumers(tenantKey);
+
+        if (StringUtils.isEmpty(config)) {
+            topicManagerService.stopAllTenantConsumers(tenantKey, existingConsumers);
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:66]: we do not need manual actions, maybe better just implement DynamicConsumerConfiguration?

Comment (sergeysenja1992) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:66]: I mean TopicConfigurationService must implements DynamicConsumerConfiguration and just call refresh consumers

Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java]: done

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    private AntPathMatcher matcher = new AntPathMatcher();
+    private ObjectMapper ymlMapper = new ObjectMapper(new YAMLFactory());
+
+    private final String configPath;
+
+    private final MessageHandler messageHandler;
+
+    private final DynamicConsumerConfigurationService dynamicConsumerConfigurationService;
+
+    @Getter
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:40]: what reason for getter?

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        refreshConfig(updatedKey, config);
+    }
+
+    @Override
+    public boolean isListeningConfiguration(String updatedKey) {
+        return matcher.match(configPath, updatedKey);
+    }
+
+    @Override
+    public void onInit(String configKey, String configValue) {
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:62]: do we need onInit?

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }
+
+    @Override
+    public List<DynamicConsumer> getDynamicConsumers(String tenantKey) {
+        if (tenantTopicConsumers.containsKey(tenantKey)) {
+            return tenantTopicConsumers.get(tenantKey);
+        }
+
+        return new ArrayList<>();
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:74]: emptyList()

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        dynamicConsumerConfigurationService.refreshDynamicConsumers(tenantKey);
+    }
+
+    private String extractTenant(final String updatedKey) {
+        return matcher.extractUriTemplateVariables(configPath, updatedKey).get(TENANT_NAME);
+    }
+
+    private Optional<TopicConsumersSpec> readSpec(String updatedKey, String config) {
+        TopicConsumersSpec spec = null;
```
Comment (sergeysenja1992) [inline xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java:100]: TopicConsumersSpec spec = ymlMapper.readValue(config, TopicConsumersSpec.class);
return Optaion.of(spec);

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java (hunk 6 above)
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java]: changed to unmodifiable getter

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java (hunk 7 above)
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java]: removed

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java (hunk 8 above)
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/TopicConfigurationService.java]: changed

File: xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class DynamicConsumerConfigurationService {
+
+    private final Map<String, Map<String, ConsumerHolder>> dynamicConsumersByTenant = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (alexmoldavsky) [reply xm-commons-topic/src/main/java/com/icthh/xm/commons/topic/service/DynamicConsumerConfigurationService.java]: moved

**Review by amedvedchuk — APPROVED** [review]

File: gradle/version.gradle [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import java.util.regex.Pattern
 
-version = "2.2.25"//detectSemVersion()
+version = "2.2.26"//detectSemVersion()
```
Comment (major13ua) [inline gradle/version.gradle:3]: After merge, please create PR to version 3.0  (java17)

**Review by major13ua — APPROVED** [review]

**Pull request merged**